azurerm/internal/services/mysql/mysql_server_key_resource.go

```diff
@@ -0,0 +1,194 @@
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/clients"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/locks"
+	keyVaultParse "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/keyvault/parse"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/mysql/parse"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/mysql/validate"
+	azSchema "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/tf/schema"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/timeouts"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func resourceArmMySQLServerKey() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmMySQLServerKeyCreateUpdate,
+		Read:   resourceArmMySQLServerKeyRead,
+		Update: resourceArmMySQLServerKeyCreateUpdate,
+		Delete: resourceArmMySQLServerKeyDelete,
+
+		Importer: azSchema.ValidateResourceIDPriorToImport(func(id string) error {
+			_, err := parse.MySQLServerKeyID(id)
+			return err
+		}),
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(30 * time.Minute),
+			Read:   schema.DefaultTimeout(5 * time.Minute),
+			Update: schema.DefaultTimeout(30 * time.Minute),
+			Delete: schema.DefaultTimeout(30 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"server_id": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validate.MysqlServerServerID,
+			},
+
+			"key_vault_key_id": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: azure.ValidateKeyVaultChildId,
+			},
+		},
+	}
+}
+
+func getMySQLServerKeyName(ctx context.Context, vaultsClient *keyvault.VaultsClient, keyVaultKeyURI string) (string, error) {
+	keyVaultKeyID, err := azure.ParseKeyVaultChildID(keyVaultKeyURI)
+	if err != nil {
+		return "", err
+	}
+	keyVaultIDRaw, err := azure.GetKeyVaultIDFromBaseUrl(ctx, vaultsClient, keyVaultKeyID.KeyVaultBaseUrl)
+	if err != nil {
+		return "", err
+	}
+	keyVaultID, err := keyVaultParse.KeyVaultID(*keyVaultIDRaw)
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("%s_%s_%s", keyVaultID.Name, keyVaultKeyID.Name, keyVaultKeyID.Version), nil
+}
+
+func resourceArmMySQLServerKeyCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	keysClient := meta.(*clients.Client).MySQL.ServerKeysClient
+	vaultsClient := meta.(*clients.Client).KeyVault.VaultsClient
+	ctx, cancel := timeouts.ForCreateUpdate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	serverID, err := parse.MysqlServerServerID(d.Get("server_id").(string))
+	if err != nil {
+		return err
+	}
+	keyVaultKeyURI := d.Get("key_vault_key_id").(string)
+	name, err := getMySQLServerKeyName(ctx, vaultsClient, keyVaultKeyURI)
```

Comment:
we could probably simplify this to only use the version of the secret id instead, since we're not using anything else?

Reply:
I am not quite sure if I get the point - I get the way we are composing the key name here from Azure CLI, they are making the key name in this pattern: vaultname_keyname_keyversion

It is curious that the service does not accept just a random name - internally they must have a pattern for this name. I wrote an email to the service team, asked about exactly what this pattern is. But the contact told me she did not know much about this. Therefore I just took the way in Azure CLI here.